File: app/src/main/java/se/accidis/fmfg/app/utils/JSONUtils.java
package se.accidis.fmfg.app.utils;

import android.text.TextUtils;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Utility methods for working with JSON.
 */
public final class JSONUtils {
	private JSONUtils() {
	}

	public static DateTime getDateTimeOrNull(JSONObject json, String key) throws JSONException {
		String str = getStringOrNull(json, key);
		if (!TextUtils.isEmpty(str)) {
			DateTimeFormatter parser = ISODateTimeFormat.dateTimeNoMillis();
			return parser.parseDateTime(str);
		} else {
			return null;
		}
	}

	public static String getStringOrNull(JSONObject json, String key) throws JSONException {
		return (json.isNull(key) ? null : json.getString(key));
	}

	public static void putDateTime(JSONObject json, String key, DateTime dateTime) throws JSONException {
		if (null != dateTime) {
			DateTimeFormatter formatter = ISODateTimeFormat.dateTimeNoMillis();
			json.put(key, dateTime.toString(formatter));
		} else {
			json.put(key, null);
		}
	}

	public static void putIfTrue(JSONObject json, String key, boolean value) throws JSONException {
		if (value) {
			json.put(key, true);
		}
	}
}

File: app/src/main/java/se/accidis/fmfg/app/utils/JSONUtils.kt
package se.accidis.fmfg.app.utils

import org.joda.time.DateTime
import org.joda.time.format.ISODateTimeFormat
import org.json.JSONObject

/**
 * Utility methods for working with JSON.
 */
object JSONUtils {
    private const val BOOLEAN_FALSE = 2
    private const val BOOLEAN_NULL = 0
    private const val BOOLEAN_TRUE = 1

    fun optDateTime(json: JSONObject, key: String): DateTime? {
        val str = json.optString(key)
        if (!str.isEmpty()) {
            val parser = ISODateTimeFormat.dateTimeNoMillis()
            return parser.parseDateTime(str)
        } else {
            return null
        }
    }

    fun putDateTime(json: JSONObject, key: String, dateTime: DateTime?) {
        if (null != dateTime) {
            val formatter = ISODateTimeFormat.dateTimeNoMillis()
            json.put(key, dateTime.toString(formatter))
        } else {
            json.put(key, null)
        }
    }

    fun optBoolean(json: JSONObject, key: String?): Boolean {
        val value = json.optInt(key, BOOLEAN_NULL)
        return value == BOOLEAN_TRUE
    }

    fun putBoolean(json: JSONObject, key: String, value: Boolean?) {
        if (null != value) {
            val intVal = (if (value) BOOLEAN_TRUE else BOOLEAN_FALSE)
            json.put(key, intVal)
        }
    }

    fun putBooleanIfTrue(json: JSONObject, key: String, value: Boolean) {
        if (value) {
            json.put(key, true)
        }
    }

    fun putOptString(json: JSONObject, key: String, value: String) {
        if (value.isNotEmpty()) {
            json.put(key, value)
        }
    }
}
